qiskit/circuit/quantumcircuit.py

```diff
@@ -2854,7 +2854,14 @@ def _assign_parameter(self, parameter: Parameter, value: ParameterValueType) ->
                     new_param = assignee.assign(parameter, value)
                     # if fully bound, validate
                     if len(new_param.parameters) == 0:
-                        instr.params[param_index] = instr.validate_parameter(new_param)
+                        if new_param._symbol_expr.is_integer and new_param.is_real():
+                            val = int(new_param)
+                        elif new_param.is_real():
+                            # Workaround symengine not supporting float(<ComplexDouble>)
+                            val = complex(new_param).real
```

Comment:
Should this `complex(self).real` form be moved into `ParameterExpression.__float__` or left here? One reason I had it here was `is_real()` was already called, and I wasn't sure which behavior was preferred in `__float__`. `float(1+1j)` is a TypeError in standard Python.

A couple other things I noticed related to this:

* symengine/sympy treat complex numbers as floating point, so something involving complex integers will end up here and evaluate to a float rather than an int even if the math involves integers that evaluate to having no imaginary part.
* The `ParameterExpression.__float__` is misleading when the value is complex because it catches the sympy/symengine error and raises an error about unbound parameters instead of the complex vs float error message.

Reply:
The `__float__` method is only meant to be implemented for types for which there is an exact conversion (unlike `__int__`, which has "casting" semantics), so it sounds like Symengine's behaviour is allowable by the spec.  If we want to do something slightly tweaked, it'd be good to unify it in `ParameterExpression.__float__` rather than here.  I think in the interest of landing this PR for a patch release today, let's look at doing that in a follow-up, though.

Bullet 1: there's no such thing as a complex integral type in Python either, so I don't think we need to worry about that (also, `ParameterExpression` is not entirely designed to represent integers).

Bullet 2: yeah, I believe it was an oversight when complex-number support was added to `ParameterExpression`.  There's issue #9187 tracking that.